crates/core/tedge_agent/src/state_repository/state.rs

```diff
@@ -1,186 +1,106 @@
 
-    async fn load(&self) -> Result<State, StateError> {
+    /// Load the latest operation, if any
+    pub async fn load(&self) -> Result<Option<T>, StateError> {
         let text = fs::read_to_string(&self.state_repo_path)
             .await
             .map_err(|e| StateError::LoadingFromFileFailed {
                 path: self.state_repo_path.as_path().into(),
                 source: e,
             })?;
 
-        let state = toml::from_str::<State>(&text).map_err(|e| StateError::FromTOMLParse {
-            path: self.state_repo_path.as_path().into(),
-            source: e,
-        })?;
-
-        Ok(state)
-    }
-
-    async fn store(&self, state: &State) -> Result<(), StateError> {
-        let toml = toml::to_string_pretty(&state)?;
-
-        // Create in path given through `config-dir` or `/etc/tedge` directory in case it does not exist yet
-        if !self.state_repo_root.exists() {
-            fs::create_dir(&self.state_repo_root).await?;
+        if text.is_empty() {
+            Ok(None)
+        } else {
+            let state = serde_json::from_str(&text).map_err(|e| StateError::InvalidJson {
+                path: self.state_repo_path.as_path().into(),
+                source: e,
+            })?;
+            Ok(Some(state))
         }
-
-        let () = atomically_write_file_async(&self.state_repo_path, toml.as_bytes()).await?;
-
-        Ok(())
     }
 
-    async fn clear(&self) -> Result<State, Self::Error> {
-        let state = State {
-            operation_id: None,
-            operation: None,
-        };
-        self.store(&state).await?;
-
-        Ok(state)
-    }
-
-    async fn update(&self, status: &StateStatus) -> Result<(), Self::Error> {
-        let mut state = self.load().await?;
-        state.operation = Some(status.to_owned());
-
-        self.store(&state).await?;
-
+    /// Store the current operation, persisting is JSON representation
+    pub async fn store(&self, state: &T) -> Result<(), StateError> {
+        let json = serde_json::to_string(state)?;
+        atomically_write_file_async(&self.state_repo_path, json.as_bytes()).await?;
         Ok(())
     }
-}
-
-impl AgentStateRepository {
-    #[cfg(test)]
-    pub fn new(tedge_root: Utf8PathBuf) -> Self {
-        Self::new_with_file_name(tedge_root, "current-operation")
-    }
-
-    pub fn new_with_file_name(tedge_root: Utf8PathBuf, file_name: &str) -> Self {
-        let mut state_repo_root = tedge_root;
-        state_repo_root.push(Utf8PathBuf::from_str(".agent").expect("infallible"));
 
-        let mut state_repo_path = state_repo_root.clone();
-        state_repo_path.push(Utf8PathBuf::from_str(file_name).expect("infallible"));
-
-        Self {
-            state_repo_path,
-            state_repo_root,
-        }
+    /// Clear the current operation by clearing the persisted file
+    pub async fn clear(&self) -> Result<(), StateError> {
+        atomically_write_file_async(&self.state_repo_path, "".as_bytes()).await?;
```

Comment:
We should be deleting the file itself here, right? Instead of just clearing the contents? Or the empty file aids with some logic? It's currently harmless as empty content is currently ignored on load. But, still...

Reply:
I tried to not change the logic of the restart command (which has not been so easy to get right) but to only change the file format to persist a context along the bare minimum info previously stored.